test/unit/conditions/condition-set.test.ts

```diff
@@ -1,75 +1,270 @@
-      ]);
-      expect(() => setWithOperatorInsteadOfComparator.validate()).toThrow(
-        'index 2 must be a Condition, got Operator instead'
-      );
-
-      const setWithConditionInsteadOfOperator = new ConditionSet([
-        cond1,
-        cond2,
-        cond1,
-      ]);
-      expect(() => setWithConditionInsteadOfOperator.validate()).toThrow(
-        'index 1 must be an Operator, got ERC721Balance instead'
-      );
-    });
+  const rpcCondition = new RpcCondition(testRpcConditionObj);
+  const timeCondition = new TimeCondition(testTimeConditionObj);
+  const compoundCondition = new CompoundCondition({
+    operator: 'and',
+    operands: [
+      testContractConditionObj,
+      testTimeConditionObj,
+      testRpcConditionObj,
+      {
+        operator: 'or',
+        operands: [testTimeConditionObj, testContractConditionObj],
+      },
+    ],
   });
 
   describe('serialization', () => {
-    it('serializes to and from json', async () => {
-      const set = new ConditionSet([
-        new ERC721Balance({
-          chain: TEST_CHAIN_ID,
-          contractAddress: TEST_CONTRACT_ADDR,
-        }),
-      ]);
-      const setJson = set.toJson();
-      expect(setJson).toBeDefined();
-      expect(setJson).toContain('chain');
-      expect(setJson).toContain(TEST_CHAIN_ID.toString());
-      expect(setJson).toContain('contractAddress');
-      expect(setJson).toContain(TEST_CONTRACT_ADDR.toString());
-
-      const setFromJson = ConditionSet.fromJSON(setJson);
-      expect(setFromJson).toBeDefined();
-      expect(setFromJson.conditions.length).toEqual(1);
-      expect(setFromJson.conditions[0]).toBeInstanceOf(Condition); // TODO: This should arguably be an ERC721Balance
+    it.each([
+      erc721BalanceCondition,
+      contractConditionNoAbi,
+      contractConditionWithAbi,
+      rpcCondition,
+      timeCondition,
+      compoundCondition,
+    ])('serializes to and from json', async (condition) => {
+      const conditionSet = new ConditionSet(condition);
+      const conditionSetJson = conditionSet.toJson();
+      expect(conditionSetJson).toBeDefined();
+      const conditionSetFromJson = ConditionSet.fromJSON(conditionSetJson);
+      expect(conditionSetFromJson).toBeDefined();
+      expect(conditionSetFromJson.equals(conditionSetFromJson)).toBeTruthy();
     });
   });
+
+  it.each([
+    // no "operator" nor "method" value
+    {
+      randoKey: 'randoValue',
+      otherKey: 'otherValue',
+    },
+    // invalid "method" and no "contractAddress"
+    {
+      method: 'doWhatIWant',
+      returnValueTest: {
+        index: 0,
+        comparator: '>',
+        value: '100',
+      },
+      chain: 5,
+    },
+    // condition with wrong method "method" and no contract address
+    {
+      ...testTimeConditionObj,
+      method: 'doWhatIWant',
+    },
+    // rpc condition (no contract address) with disallowed method
+    {
+      ...testRpcConditionObj,
+      method: 'isPolicyActive',
+    },
```

Comment:
What do you think about using permutations here to exhaustively test for every possible incorrect variant? Nothing from with the current implementation, I'm just interested in your thoughts.

Reply:
I'm intrigued - could you elaborate a bit more on what you are thinking? Just want to make sure I understand.

The goal for this test was to test the `ConditionSet.fromObj(...)` branching logic for invalid data.

Reply:
We could generate a set of all possible permutations of the form `<field>: <value>`, where both `field` and `value` may contain valid and invalid values. From this set we remove permutations we know to be correct. Others should fail the validation, and so the test cases would be successful.

I think without a property-based testing library this is tiresome to implement. I did consider this type of testing in other places on different occasions.